### PR TITLE
Adopt BBE to 0.990 syntax

### PR DESCRIPTION
**Pull request by ChinthakaSenanayaka**
## Automation tests
 - Unit tests 
   > Code coverage information N/A
 - Integration tests
   > Details about the test cases and coverage N/A

## Security checks
 - Followed secure coding standards in http://wso2.com/technical-reports/wso2-secure-engineering-guidelines? N/A
 - Ran FindSecurityBugs plugin and verified report? N/A
 - Confirmed that this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets? N/A

### CODE REVIEW

File: examples/any-type/any_type.bal [hunk 1]
```diff
@@ -12,8 +12,13 @@ public function main() {
     io:println(a);
 
     // First, the variable of the `any` type needs to be cast to the required type (`int` in this example) as shown here.
-    int intVal = check <int>a;
-    io:println(intVal + 10);
+    int|error intVal = trap <int>a;
+    if (intVal is int) {
+        io:println(intVal + 10);
+    } else {
+        // Runtime value is casted to correct type since Ballerina runtime can infer the correct type to error.
```
Comment (pubudu91) [inline examples/any-type/any_type.bal]: Not sure if "casted" is the correct term to use here. @SupunS thoughts?

File: examples/error-handling/error_handling.description [hunk 2]
```diff
@@ -1,2 +1,2 @@
-// In Ballerina, error handling can be performed either by returning a value of type `error`, or by throwing an exception.
+// In Ballerina, error handling can be performed either by returning a value of type `error`, or by panicking an error.
```
Comment (pubudu91) [inline examples/error-handling/error_handling.description]: I think we may have to rephrase and elaborate a bit more considering that errors have changed significantly

File: examples/function-pointers/function_pointers.bal [hunk 3]
```diff
@@ -2,15 +2,22 @@ import ballerina/io;
...
 function test(int x, string s) returns (float) {
-    int y = check <int> s;
-    float f = x * 1.0 * y;
+    int|error y = int.create(s);
+    float f = 0.0;
+
+    if (y is int) {
+        f = x * 1.0 * y;
+    } else {
+        // Runtime value is casted to correct type since Ballerina runtime can infer the correct type to error.
```
Comment (pubudu91) [inline examples/function-pointers/function_pointers.bal]: Again, not sure if "casted" is the correct term

File: examples/function-pointers/function_pointers.bal [hunk 4]
```diff
@@ -2,15 +2,22 @@ import ballerina/io;
...
+        panic y;
+    }
     return f;
 }
 
-// Function pointer as a parameter.
+// Function pointer as a parameter. Use `call` function to invoke the function using the function pointer.
```
Comment (pubudu91) [inline examples/function-pointers/function_pointers.bal]: `call` -> `call()`

File: examples/elvis-operator/elvis_operator.bal [hunk 5]
```diff
@@ -1,16 +1,18 @@
...
+    string|() x = ();
 
-    // If you need to get the string value of x, and if the value of x is `null`, you may want
-    // to assign a known value. This is how you can do it via Match
-    // Expression.
-    string matchExprOutput = x is string ? s : "value is null";
-
-    io:println("The output from match expression: ", matchExprOutput);
+    // If you need to get the `string` value of x, and if the value of x is `nil`, you may want
+    // to assign a known value. This is how you can do it via type-guard (other word `is` check)
```
Comment (samgnaniah) [inline examples/elvis-operator/elvis_operator.bal]: Change to "This is how you can do it via type-guard (which is also known as `is` check)."

File: examples/panic/panic.description [hunk 6]
```diff
@@ -1,8 +1,8 @@
-// In Ballerina, throwing an error means something unexpected has occured
-// (something which should not have occurred during the normal operation). A thrown error causes
+// In Ballerina, panicking an error means something unexpected has occured
```
Comment (pubudu91) [inline examples/panic/panic.description]: Shall we refrain from using terms like "panicking an error", "panicked error" and simply refer to it as a panic? e.g., https://gobyexample.com/panic

File: examples/panic/panic.bal [hunk 7]
```diff
@@ -0,0 +1,31 @@
+import ballerina/io;
+
+type Record record {
+    int id;
+    string name;
+};
+
+// This example demonstrates panic an error. The next example demonstrates how to catch the panicked errors.
```
Comment (samgnaniah) [inline examples/panic/panic.bal]: What do we mean by "panic an error" here?

Also we shouldn't refer to the next example as that might not be straightforward navigation. It's better if we talk only about this example here and give a separate intro to the other example.

File: examples/record-optional-fields/record_optional_fields.bal [hunk 8]
```diff
@@ -5,8 +5,8 @@ type Gender "male"|"female";
 type Person record {
     // This is a required field with an explicit default value specified.
     string fname = "default";
-    // This is a required field without an explicit default value. However, `string` has an implicit initial value (empty string).
-    // Therefore, the compiler will not complain if we omit this field when creating the record.
+    // This is a required field without an explicit default value.
+    // The compiler will not complain if we omit this field when creating the record.
```
Comment (pubudu91) [inline examples/record-optional-fields/record_optional_fields.bal]: This should now be "Omitting this field when creating the record will result in a compile error."

File: examples/panic/panic.description [hunk 9]
```diff
@@ -1,8 +1,8 @@
...
+// (something which should not have occurred during the normal operation). A panicked error causes
```
Comment (samgnaniah) [inline examples/panic/panic.description]: Change "which" to "that".

File: examples/record-optional-fields/record_optional_fields.bal [hunk 10]
```diff
@@ -17,7 +17,7 @@ type Person record {
 public function main() {
     // The `gender` must be provided when creating the record since `Gender` is not a type with an implicit initial value
     // and an explicit default value is not specified for it in the record descriptor.
-    Person p = {gender: "male"};
```
Comment (pubudu91) [inline examples/record-optional-fields/record_optional_fields.bal]: The description here must also change now. Maybe to something like the following: "The `gender` and `lname` fields must be provided when creating the record since explicit default values are not specified for these in the record descriptor."

File: examples/trap-error/trap_error.bal [hunk 11]
```diff
@@ -0,0 +1,23 @@
+import ballerina/io;
+
+public function main() {
+    int|error result;
+    io:println("Start dividing numbers");
+
+    // An error is panicked (or thrown) when the `divideNumbers` function is executed.
+    // `trap` will catch the error which occurred in the `divideNumbers` function and assgins the `result` variable.
```
Comment (pubudu91) [inline examples/trap-error/trap_error.bal]: Same comment as above regarding panics

Comment (pubudu91) [reply examples/trap-error/trap_error.bal]: `divideNumbers` -> `divideNumbers()`
Typo: assgins -> assigns

Comment (samgnaniah) [inline examples/trap-error/trap_error.bal]: Change to "`trap` catches the error that..."

Normally we use the word "that" in this context. "Which" is used only if you are adding additional information to the sentence and normally there is a comma before the word "which".

Also avoid future tense wherever possible. Normally if you see the word "will" in the sentence it is indicative of future tense.

File: examples/panic/panic.bal (hunk 7 above)
Comment (pubudu91) [reply examples/panic/panic.bal]: I think we shouldn't be referring to as "panicking an error". See https://gobyexample.com/panic for example

File: examples/trap-error/trap_error.description [hunk 12]
```diff
@@ -0,0 +1 @@
+// The `trap` keyword will catch an error assign it to the variable.
```
Comment (samgnaniah) [inline examples/trap-error/trap_error.description]: Change to "The `trap` keyword catches an error and assigns it to the variable."

Can we expand more on why it is useful etc.?

File: examples/any-type/any_type.description [hunk 13]
```diff
@@ -1,4 +1,4 @@
-// The `any` data type is the root of the Ballerina data types. It can represent a variable of any data type.
+// The `any` data type is the root of the Ballerina data types. It can represent a variable of any type excluding error type.
```
Comment (pubudu91) [inline examples/any-type/any_type.description]: Need to rephrase this as well. It's misleading to say `any` is the root of data types. Shall we rephrase this to something like "The `any` type is the root of all the types except `error` in Ballerina." ?

File: examples/check/check.bal [hunk 14]
```diff
@@ -12,12 +12,14 @@ type Address record {
...
 function getAddress(Person p) returns (Address|error) {
     // If the address does not exist, return an error.
-    match (p.address) {
-        Address add => { return add;}
-        () => {
-            error addNotFoundErr = error("address not found");
-            return addNotFoundErr;
-        }
+    // For type-guard to work, only simple variable references should be used. therefore, `p.address` is assigned to a
+    // variable.
```
Comment (pubudu91) [inline examples/check/check.bal]: Shall we make this explanation a bit more comprehensive? i.e., the difference between type guard and type test and maybe demonstrate both cases

File: examples/check/check.bal [hunk 15]
```diff
@@ -35,16 +37,19 @@ function validateAddress(Person person) returns (boolean|error) {
 function validateAddressAgain(Person person) returns boolean {
     //The enclosing function's return type does not include `error` as an alternative (it allows only boolean).
     //So, if the "getAddress(person)!city" expression evaluates an error, 
-    //the `check` expression throws the resulted error.
+    //the `check` expression panics the resulted error.
```
Comment (pubudu91) [inline examples/check/check.bal]: Again, improper use of the term panic

File: examples/elvis-operator/elvis_operator.bal [hunk 16]
```diff
@@ -1,16 +1,18 @@
...
-    // If you need to get the string value of x, and if the value of x is `null`, you may want
-    // to assign a known value. This is how you can do it via Match
-    // Expression.
-    string matchExprOutput = x is string ? s : "value is null";
-
-    io:println("The output from match expression: ", matchExprOutput);
+    // If you need to get the `string` value of x, and if the value of x is `nil`, you may want
+    // to assign a known value. This is how you can do it via type-guard (also known as `is` check)
+    // expression.
+    if (x is string) {
```
Comment (pubudu91) [inline examples/elvis-operator/elvis_operator.bal]: Any particular reason to change this to an if-else?

File: examples/foreach/foreach.bal [hunk 17]
```diff
@@ -21,21 +21,22 @@ public function main() {
...
+            io:println("non-string value: ", j);
         }
     }
 
     io:println("\nIterating over a json array:-");
     // To Iterate over a JSON array, you need to first cast it into an array of json (`json[]`).
-    json[] colors = check <json[]>apple.colors;
-    foreach i, j in colors {
-        io:println("color ", i, ": ", j);
+    json[]|error colors = trap <json[]>apple.colors;
```
Comment (pubudu91) [inline examples/foreach/foreach.bal]: The description for this need to be updated as well. To something like "To iterate over a JSON array, first you need to assert that it is an array of `json` (`json[]`)."

Comment (pubudu91) [reply examples/foreach/foreach.bal]: And shall we get rid of the trap here? Then we don't need the type test below.

File: examples/hello-world-service/hello_world_service.bal [hunk 18]
```diff
@@ -2,14 +2,10 @@ import ballerina/http;
 import ballerina/log;
 
 // By default, Ballerina exposes a service via HTTP/1.1.
-service<http:Service> hello bind { port: 9090 } {
+service hello on new http:Server(9090) {
```
Comment (pubudu91) [inline examples/hello-world-service/hello_world_service.bal]: This is now `http:Listener` right?

File: examples/length/length.bal [hunk 19]
```diff
@@ -5,8 +5,8 @@ public function main() {
     int[] intArray = [1, 2, 3];
     int length;
 
-    // This prints the length of the integer array created. The length is calculated using the `lengthof` unary operator.
-    length = lengthof intArray;
+    // This prints the length of the integer array created. The length is calculated using the `length` function.
```
Comment (pubudu91) [inline examples/length/length.bal]: length -> length()

File: examples/length/length.description [hunk 20]
```diff
@@ -0,0 +1 @@
+// Ballerina supports the `length` function that returns the length/size of the given value.
```
Comment (pubudu91) [inline examples/length/length.description]: We need to explain a bit more about what types of values support this function. Otherwise it'll give the wrong impression that this can be called on any type of value.

File: examples/record-optional-fields/record_optional_fields.bal (hunk 10 above)
Comment (ChinthakaSenanayaka) [reply examples/record-optional-fields/record_optional_fields.bal]: I think gender is different case than lname. Thus, explained separately.

File: examples/any-type/any_type.description (hunk 13 above)
Comment (ChinthakaSenanayaka) [reply examples/any-type/any_type.description]: Kept this for till the definition gets updated in spec

File: examples/record-optional-fields/record_optional_fields.bal (hunk 10 above)
Comment (pubudu91) [reply examples/record-optional-fields/record_optional_fields.bal]: No, that doesn't apply now. Previously, `lname` had an implicit initial value since it is a string, making it an optional field in essence. But now, since we don't have implicit initial values, `lname` is also a required field.

File: examples/any-data-type/any_data_type.bal [hunk 21]
```diff
@@ -0,0 +1,29 @@
...
+    string name = "cat";
+    return name;
+}
+
+public function main() {
+    // In this example, the variable named `a` of the `anydata` type holds an `int` value.
+    anydata a = 5;
+    io:println(a);
+
+    // First, the variable of the `anydata` type needs to be cast to the required type (`int` in this example) as shown here.
```
Comment (MaryamZi) [inline examples/any-data-type/any_data_type.bal]: How about saying
```suggestion
    // First, the variable of type `anydata` needs to be asserted to be of the required type (`int` in this example) 
```

I think we should be using the word "assert" instead of "cast" now.

File: examples/any-data-type/any_data_type.bal [hunk 22]
```diff
@@ -0,0 +1,29 @@
...
+    int|error intVal = trap <int>a;
+    if (intVal is int) {
+        io:println(intVal + 10);
+    } else {
+        // Runtime value is cast to correct type since Ballerina runtime can infer the correct type to error.
+        io:println("Error occurred: " + intVal.reason());
+    }
+
+    // In Ballerina, a variable of the `anydata` type can hold values of any data type.
```
Comment (MaryamZi) [inline examples/any-data-type/any_data_type.bal]: How about something like

```suggestion
    // A variable of type `anydata` can hold any value of an `anydata` compatible type.
```

File: examples/any-data-type/any_data_type.out [hunk 23]
```diff
@@ -0,0 +1,5 @@
+$ ballerina run any_data_type.bal
+5
+15
+[1, 3, 5, 6]
+cat
```
Comment (MaryamZi) [inline examples/any-data-type/any_data_type.out]: needs a new line.

File: examples/any-type/any_type.bal [hunk 24]
```diff
@@ -1,24 +1,29 @@
...
-    // In Ballerina, a variable of the `any` type can hold values of any data type.
-    int[] ia = [1, 3, 5, 6];
-    any ar = ia;
-    io:println(ar);
+public function main() {
+    int a = 25;
+    io:println(a);
 
-    io:println(getValue());
+    any p = new Person(age=a);
```
Comment (MaryamZi) [inline examples/any-type/any_type.bal]:
```suggestion
    any p = new Person(age = a);
```

File: examples/any-data-type/any_data_type.description [hunk 25]
```diff
@@ -0,0 +1,5 @@
+// The `anydata` type is a sub set of type `any` of the Ballerina data types. `anydata` is an union type which consists
+// boolean | int | float | decimal | string | xml | table | anydata[] | map<anydata> types.
```
Comment (pubudu91) [inline examples/any-data-type/any_data_type.description]: `nil` should be here as well. 
```suggestion
// `boolean | int | float | decimal | string | xml | table | anydata[] | map<anydata> | ()` types.
```
Also shall we rephrase this to "The `anydata` type is defined by the following union: `boolean | int | float | decimal | string | xml | table | anydata[] | map<anydata> | ()`"

File: examples/record-optional-fields/record_optional_fields.bal (hunk 10 above)
Comment (ChinthakaSenanayaka) [reply examples/record-optional-fields/record_optional_fields.bal]: done

File: examples/length/length.description (hunk 20 above)
Comment (ChinthakaSenanayaka) [reply examples/length/length.description]: done

File: examples/length/length.bal (hunk 19 above)
Comment (ChinthakaSenanayaka) [reply examples/length/length.bal]: done

File: examples/function-pointers/function_pointers.bal (hunk 4 above)
Comment (ChinthakaSenanayaka) [reply examples/function-pointers/function_pointers.bal]: done

File: examples/foreach/foreach.bal (hunk 17 above)
Comment (ChinthakaSenanayaka) [reply examples/foreach/foreach.bal]: done

File: examples/check/check.bal (hunk 14 above)
Comment (ChinthakaSenanayaka) [reply examples/check/check.bal]: done

File: examples/any-type/any_type.bal (hunk 24 above)
Comment (ChinthakaSenanayaka) [reply examples/any-type/any_type.bal]: done

File: examples/any-data-type/any_data_type.out (hunk 23 above)
Comment (ChinthakaSenanayaka) [reply examples/any-data-type/any_data_type.out]: done

File: examples/any-data-type/any_data_type.description (hunk 25 above)
Comment (ChinthakaSenanayaka) [reply examples/any-data-type/any_data_type.description]: done

File: examples/any-data-type/any_data_type.bal (hunk 22 above)
Comment (ChinthakaSenanayaka) [reply examples/any-data-type/any_data_type.bal]: done

File: examples/any-data-type/any_data_type.bal [hunk 26]
```diff
@@ -0,0 +1,29 @@
+import ballerina/io;
```
[the remark this answers is not in the document]
Comment (ChinthakaSenanayaka) [reply examples/any-data-type/any_data_type.bal]: done

File: examples/any-data-type/any_data_type.bal [hunk 27]
```diff
@@ -0,0 +1,29 @@
...
+
+// This function returns a value of the `anydata` type.
+function getValue() returns (anydata) {
```
[the remark this answers is not in the document]
Comment (ChinthakaSenanayaka) [reply examples/any-data-type/any_data_type.bal]: done

File: examples/any-data-type/any_data_type.bal (hunk 21 above)
Comment (ChinthakaSenanayaka) [reply examples/any-data-type/any_data_type.bal]: done
